### PR TITLE
SPACE_TO_BATCH_ND: update output tensor shape

### LINKED ISSUE #2319: Dilated Convolutional Layers Produce Wrong Output
Issue: Models containing dilated convolutional layers produce incorrect output. This is true for 1D and 2D convolutional layers, all tested dilation_rate sizes, and for both fp32 and int8 quantized models. This issue has been brought up in #1646, #1629, #1231 which had no response nor resolution and closed to due being stale.

Reproduce: Running the following script will show that TF and TFLite produce correct results while TFLM does not.

```python

import numpy as np
import tensorflow as tf
import tflite_micro as tflm

INPUT_SHAPE = (8, 1)
DILATION_RATE = 2  # <- If 1 then assertions pass

# Create example dilated convolutional model
inputs = tf.keras.layers.Input(shape=INPUT_SHAPE)
y = inputs
y = tf.keras.layers.Convolution1D(1, 3, dilation_rate=DILATION_RATE, padding='same')(y)
model = tf.keras.Model(inputs=inputs, outputs=y)


# Convert to TF Lite model
converter = tf.lite.TFLiteConverter.from_keras_model(model=model)
tflite_model = converter.convert()

# Create TFLite and TFLM interpreter
tfl_interpreter = tf.lite.Interpreter(model_content=tflite_model)
tfl_interpreter.allocate_tensors()

tflm_interpreter = tflm.runtime.Interpreter.from_bytes(tflite_model)

# Create input tensor of 1s
x = np.ones((1,)+INPUT_SHAPE, dtype=np.float32)

# Invoke TF, TFLite and TFLM models using same input
y_tf = model(x)

tfl_interpreter.set_tensor(tfl_interpreter.get_input_details()[0]["index"], x)
tfl_interpreter.invoke()
y_tfl = tfl_interpreter.get_tensor(tfl_interpreter.get_output_details()[0]["index"])

tflm_interpreter.set_input(x, 0)
tflm_interpreter.invoke()
y_tflm = tflm_interpreter.get_output(0)

# Below passes
np.testing.assert_almost_equal(y_tf, y_tfl, decimal=3, err_msg="TF and TFLite output do not match.")

# Below fails
np.testing.assert_almost_equal(y_tf, y_tflm, decimal=3, err_msg="TF and TFLM output do not match.")


```

### LINKED ISSUE #2332: SPACE_TO_BATCH_ND: update output tensor shape
@tensorflow/micro 

The SPACE_TO_BATCH_ND kernel needs to update the output tensor shape to provide correct output results.

Need to add additional tests from TfLite.

### PR DESCRIPTION
**Pull request by ddavis-2015**
@tensorflow/micro

Update the output tensor shape during prepare phase when the computed shape does not match the shape in the flatbuffer.

Added additional tests from TfLite.

See #2319 for additional details.

Resolves [b/313360569](https://issuetracker.google.com/313360569)

bug=fixes #2332 fixes #2319 #1646 #1629 #1231

### CODE REVIEW

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc [hunk 1]
```diff
@@ -44,6 +46,67 @@ void* Init(TfLiteContext* context, const char* buffer, size_t length) {
...
+  output_shape.SetDim(input_dims->size - 1,
+                      input_dims->data[input_dims->size - 1]);
+
+  // check if need to relocate output tensor dims
+  if (output_shape == old_output_shape) {
+    return kTfLiteOk;
+  } else if (output_shape.FlatSize() > old_output_shape.FlatSize() &&
+             output->data.data != nullptr) {
+    MicroPrintf(
+        "SPACE_TO_BATCH_ND: resizing flatbuffer tensor data is not supported");
```
Comment (rascani) [inline tensorflow/lite/micro/kernels/space_to_batch_nd.cc:94]: I think it would be more accurate to say that "increasing tensor size is not supported" as we are supporting downward resizing.

**Review by rascani — COMMENTED** [review]
LGTM, just a couple comment nits.

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc (hunk 1 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/space_to_batch_nd.cc:94]: SPACE_TO_BATCH_ND can increase the output tensor data buffer size depending on the block_shape and padding.  In cases where the data buffer has not yet been allocated (MicroInterpreter) this is ok, as we just set a larger shape than what was in the flatbuffer.  For the KernelRunner however, we must give this error as the tensor data buffer is always pre-allocated prior to the prepare phase.

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc [hunk 2]
```diff
@@ -44,6 +46,67 @@ void* Init(TfLiteContext* context, const char* buffer, size_t length) {
...
+  // Block_shape should be a 1D tensor with dimension [spatial_dims_num].
+  TF_LITE_ENSURE_EQ(context, NumDimensions(block_shape), 1);
+  TF_LITE_ENSURE_EQ(context, block_shape->dims->data[0], spatial_dims_num);
+  // Padding should be a 2D tensor with dimension [spatial_dims_num, 2].
+  TF_LITE_ENSURE_EQ(context, NumDimensions(padding), 2);
+  TF_LITE_ENSURE_EQ(context, padding->dims->data[0], spatial_dims_num);
+  TF_LITE_ENSURE_EQ(context, padding->dims->data[1], 2);
+
+  RuntimeShape output_shape =
+      GetTensorShape(input);  // yes, copy from input tensor
```
[the remark this answers is not in the document]
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/space_to_batch_nd.cc]: done

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc (hunk 1 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/space_to_batch_nd.cc:95]: how about "illegal tensor size change"?

**Review by rascani — APPROVED** [review]

**Comment by ddavis-2015** [comment]
This PR is superseded by PR #2356
